[wip] Fix phi placement

o- opened the pull request.

o- committed:
use the standard phi placement algo

JanJecmen commented on rir/src/compiler/analysis/verifier.cpp:
```diff
@@ -214,6 +225,21 @@ class TheVerifier {
+                std::unordered_set<BB*> inp;
+                for (auto in : cfg(bb->owner).immediatePredecessors(bb))
+                    inp.insert(in);
+                phi->eachArg([&](BB* bb, Value*) { inp.erase(inp.find(bb)); });
```
this can backfire if bb is not found:
> The iterator pos must be valid and dereferenceable. Thus the end() iterator (which is valid, but is not dereferenceable) cannot be used as a value for pos.

mhyee commented on rir/src/compiler/pir_tests.cpp:
```diff
@@ -394,7 +394,106 @@ bool testPir2Rir(const std::string& name, const std::string& fun,
+            entry = e;
+            nextBBId = s;
+        }
+        ~MockCode() { entry = new MockBB; }
```
???

o- replied:
:) don't ask. `~Code` insists on deleting something

mhyee replied:
Now that I looked at the [code](https://github.com/reactorlabs/rir/blob/master/rir/src/compiler/pir/code.cpp#L28) it makes sense. Please add a comment, because by itself, this line is very 😖 .

mhyee commented on rir/src/compiler/pir_tests.cpp:
```diff
@@ -394,7 +394,106 @@ bool testPir2Rir(const std::string& name, const std::string& fun,
+};
+MockBB::MockCode MockBB::code = MockCode(nullptr, 0);
+
+bool testCfg() {
```
A bit of ASCII art here would make it easier to follow the examples.

mhyee commented:
Also, I gave up trying to understand what `scope_resolution` and `phi_placement` are doing.

o- commented:
@mhyee It's not that complicated... scope analysis tracks all `stvars`. this allows you to statically approximate ldvars. the scope analysis will tell you all possible values for a particular load. from that set of inputs, you need to then construct ssa variables, if possible.

o- committed:
fix bb renumbering

we are using the dominance tree visitor to renumber bbs. this visitor
is now relying on the DominanceGraph, which unfortunately assumes
stable ids. So we need to cache the dominance graph before renumbering.

charig commented on rir/src/utils/Set.h:
```diff
@@ -0,0 +1,65 @@
+#ifndef RIR_SET_H
```
Why???

o- replied:
* std::set :  needs "operator<" implemented and is generally big overhead
* std::unordered_set :  needs hash implemented, easy to get wrong, big overhead for small sets
* SmallSet, just needs `operator==` and is faster and less memory overhead for small sets

o- replied:
just wanted something simpler...
